src/charm.py

```diff
@@ -699,39 +702,113 @@ def _activate_charm(self) -> None:
             self._start_service()
             self.model.unit.status = ActiveStatus()
 
-    def _on_add_admin_user_action(self, event: ActionEvent) -> None:
-        """Add a new admin user to Discourse.
+    def _on_promote_user_action(self, event: ActionEvent) -> None:
+        """Promote a user to a specific trust level.
 
         Args:
-            event: Event triggering the add_admin_user action.
+            event: Event triggering the promote_user action.
         """
-        email = event.params["email"]
-        password = event.params["password"]
         container = self.unit.get_container(CONTAINER_NAME)
-        if container.can_connect():
-            process = container.exec(
-                [
-                    os.path.join(DISCOURSE_PATH, "bin/bundle"),
-                    "exec",
-                    "rake",
-                    "admin:create",
-                ],
-                stdin=f"{email}\n{password}\n{password}\nY\n",
-                working_dir=DISCOURSE_PATH,
-                user=CONTAINER_APP_USERNAME,
-                environment=self._create_discourse_environment_settings(),
-                timeout=60,
+        if not container.can_connect():
+            event.fail("Unable to connect to container, container is not ready")
+            return
+
+        email = event.params["email"]
+
+        user_exists = container.exec(
+            [os.path.join(DISCOURSE_PATH, "bin/bundle"), "exec", "rake", f"users:exists[{email}]"],
+            working_dir=DISCOURSE_PATH,
+            user=CONTAINER_APP_USERNAME,
+            environment=self._create_discourse_environment_settings(),
+        )
+        try:
+            user_exists.wait_output()
+        except ExecError:
+            event.fail(f"User with email {email} does not exist")
+            return
+
+        process = container.exec(
+            [
+                os.path.join(DISCOURSE_PATH, "bin/bundle"),
+                "exec",
+                "rake",
+                "admin:create",
+            ],
+            stdin=f"{email}\nn\nY\n",
+            working_dir=DISCOURSE_PATH,
+            user=CONTAINER_APP_USERNAME,
+            environment=self._create_discourse_environment_settings(),
+            timeout=60,
+        )
+        try:
+            process.wait_output()
+            event.set_results({"user": email})
+        except ExecError as ex:
+            event.fail(
+                f"Failed to make user with email {email} an admin: {ex.stdout}"  # type: ignore
             )
+
+    def _on_create_user_action(self, event: ActionEvent, skip_validation: bool = False) -> None:
```

Comment:
Where is `skip_validation` used?

Reply:
When we mock in the unit test, all container exec calls fire without failing, even if the rake task fails. 

In this test, we rely on the `users:exists` rake call to fail, to proceed with the user creation. Because this always passes when mocked, the function exits early and the rake command to create the user never occurs. Thus, we need to have some option of skipping validation in order for the unit test to verify the rake command works.

The `create_user` action is tested fully, with more rigour, in the integration test that accompanies it. Unfortunately these rake methods cannot be tested so simply with unit tests

Reply:
Can't you just patch the `.wait_output` return value instead? It'd be better not to have source code that's only used for testing

Reply:
patch added